### PR TITLE
Don't bottleneck Wave intrinsics through `WaveMask*` for spirv.

### LINKED ISSUE #4097: Incorrect WaveMatch result with SPIR/V
According to the "spec" the `WaveMatch` intrinsic should return `0` for any lane that is inactive or beyond `WaveGetLaneCount`. This is what I see when running a test shader compiled to HLSL. When generating SPIR/V, at least on NVIDIA the returned mask `uint4` repeats the value of `.x` across all other components instead of them being `0`.

from https://microsoft.github.io/DirectX-Specs/d3d/HLSL_ShaderModel6_5.html :

> The return value is a uint4 representing a 128b bit mask which identifies lanes in the current wave matching current lane’s value of val. Bits in the mask corresponding to inactive lanes, or at positions beyond current implementation’s wave width, will contribute 0’s. Bits in the mask corresponding to active lanes which match the value of val in the current lane will be set to 1. The bit in the mask corresponding to the current lane will always be set to 1.

### PR DESCRIPTION
**Pull request by csyonghe**
Fixes #4097.

### CODE REVIEW

File: source/slang/hlsl.meta.slang [hunk 1]
```diff
@@ -12305,6 +12359,14 @@ uint4 WaveMatch(T value)
     __target_switch
     {
     case hlsl: __intrinsic_asm "WaveMatch";
+    case glsl: __intrinsic_asm "subgroupPartitionNV($0)";
```
Comment (sriramm-nv) [inline source/slang/hlsl.meta.slang:12362]: is subgroupPartitionNV part of HLSL as well?

Comment (csyonghe) [reply source/slang/hlsl.meta.slang:12362]: This is implementing the intrinsic for the `case glsl` case, so it is for GLSL.

Comment (sriramm-nv) [reply source/slang/hlsl.meta.slang:12362]: oh i thought HLSL was pass through case.

**Comment by sriramm-nv** [comment]
looks good to me except for that one line.

File: source/slang/hlsl.meta.slang (hunk 1 above)
Comment (csyonghe) [reply source/slang/hlsl.meta.slang:12362]: __intrinsic_asm is like a `return`, and there is no passthrough for __target_switch.

File: source/slang/hlsl.meta.slang [hunk 2]
```diff
@@ -12329,6 +12399,14 @@ uint4 WaveMatch(matrix<T,N,M> value)
     __target_switch
     {
     case hlsl: __intrinsic_asm "WaveMatch";
+    case glsl:
+    case cuda:
+    case spirv:
+        uint4 result = uint4(0xFFFFFFFF);
+        [ForceUnroll]
+        for (int i = 0; i < N; i++)
+            result &= WaveMatch(value[0]);
```
Comment (jkwak-work) [inline source/slang/hlsl.meta.slang:12408]: Shouldn't this be like this?
```
result &= WaveMatch(value[i]);
```

Comment (csyonghe) [reply source/slang/hlsl.meta.slang:12408]: You are right. Thanks for catching this, will fix in a new PR.

**Pull request merged**